### ISSUE TITLE
Add "raw" flag equivalent for JQ

### ISSUE THREAD

**Issue by LorisFriedel** [bitfield/script#125]
Hello! Thank you for the amazing work :)

I would like to suggest the possibility to output a string using JQ but with the equivalent of "-r" or "--raw" we can find on the official jq command: this allows to get a clean output for a single field for example, instead of having to sanitize it by doing some triming.

What do you think?

Thanks!

**Comment by bitfield** [comment]
Nice suggestion, @LorisFriedel! Can you give a brief example of a program like this, showing how you'd like this behaviour to work?

**Comment by LorisFriedel** [comment]
For example, I would ideally do this to find a single element in a json object and use it directly:
```golang
nodeName, err := script.
		Exec("kubectl get pod myPod -o json").
		JQRaw(".spec.nodeName").
		String()

// output would be: xxx-ffkch-something-e32as-someregion-gsbn4
```

Today, that's what I need to do:
```golang
nodeName, err := script.
		Exec("kubectl get pod myPod -o json").
		JQRaw(".spec.nodeName").
		String()

// output is: "xxx-ffkch-something-e32as-someregion-gsbn4"\n

nodeName = strings.Trim(strings.TrimSpace(nodeName), "\"") // extra step to sanitze the output
// output is then: xxx-ffkch-something-e32as-someregion-gsbn4
```

**Comment by bitfield** [comment]
Maybe this could be done with a more general-purpose trim method—something like `Trim`?

```go
nodeName, err := script.
		Exec("kubectl get pod myPod -o json").
		JQ(".spec.nodeName").
		Trim().
		String()
// output would be: xxx-ffkch-something-e32as-someregion-gsbn4
```

**Comment by LorisFriedel** [comment]
This could be, but the trim would need to strip spaces AND quotes, and that would make it a weird "Trim" method I guess, WDYT?

**Comment by bitfield** [comment]
Maybe there's a better name for this...

**Comment by LorisFriedel** [comment]
Maybe just a `Trim(cutset string)` and `TrimSpace()` functions could be perfectly suited here, easy to implement and easy to understand as they match their `strings.Trim` and `strings.TrimSpace` equivalent! WDYT?

**Comment by bitfield** [comment]
I'd like to gather a bit more data on exactly how people are using `JQ` within `script` programs, so that we can come up with an API which fits well in that context. So if you're reading this issue and you have a program that uses `jq` queries, please post it in a comment!

**Comment by kwilczynski** [comment]
Hi @bitfield,

> I'd like to gather a bit more data on exactly how people are using `JQ` within `script` programs, so that we can come up with an API which fits well in that context. So if you're reading this issue and you have a program that uses `jq` queries, please post it in a comment!

I cannot speak for every jq user out there however, using **jq** with the `-r` command-line switch is definitely a thing, as otherwise, you often have to resort to some **sed**, **grep** or **awk** invocation to remove the unwanted quoting and/or newlines.

**Comment by bitfield** [comment]
> using jq with the -r command-line switch is definitely a thing

I believe you, but we need a realistic example to use as the input to the design process.

**Comment by kwilczynski** [comment]
Hi @bitfield,

> > using jq with the -r command-line switch is definitely a thing
> 
> I believe you, but we need a realistic example to use as the input to the design process.

Thinking about it some more... I wouldn't do it.

Adding some convenience functions like **Trim()** or **Strip()**, and also **Tr()** might be a better alternative - new API could be easily reused somewhere else, other than removing bloat from jq's output.

**Comment by LorisFriedel** [comment]
Anything like Trim(), Strip() and such would save a lot a painful code handling those polluted outputs indeed!

**Comment by bitfield** [comment]
Is there any example you can think of where a 'trim spaces and quotes' method would be used with anything _other_ than `JQ`?

**Comment by LorisFriedel** [comment]
For the "trim quote" part, not really, but for the "trim space" part I guess everywhere where `Fprintln()` function is used in the library could be a good candidate, as when getting the output of something you generally don't want additional whitespace that serves no purposes!

**Comment by bitfield** [comment]
> An example use case (aside from using it for JQ): working with a result of a process spawned with Exec() to clean up output from the external process.

The problem I see is that a `Trim` method specialised for trimming `jq` output most likely wouldn't be much use for anything else.

Conversely, a `Trim` method that's configurable enough to trim both `jq` data and a variety of other cases would probably require an inconvenient amount of paperwork.

It's possible that we could get the best of both worlds, by providing a `Trim` method that performs general-purpose substitution, like `tr`, and *also* a "raw mode" for `JQ` that could use `Trim` under the hood.

**Comment by kwilczynski** [comment]
Hi @bitfield,

> > An example use case (aside from using it for JQ): working with a result of a process spawned with Exec() to clean up output from the external process.
> 
> The problem I see is that a `Trim` method specialised for trimming `jq` output most likely wouldn't be much use for anything else.

Not if we make it generic. I can see it being useful as a handy function that can be used to rid something in the pipeline, so to speak, of whitespaces and such.

> Conversely, a `Trim` method that's configurable enough to trim both `jq` data and a variety of other cases would probably require an inconvenient amount of paperwork.

What if we kept it simple? Like [String#strip](https://docs.ruby-lang.org/en/master/String.html#method-i-strip) as an example from Ruby.
 
> It's possible that we could get the best of both worlds by providing a `Trim` method that performs general-purpose substitution, like `tr`, and _also_ a "raw mode" for `JQ` that could use `Trim` under the hood.

Or... **Tr()** and **Strip()**, simply.

Users of **JQ** can then use either depending on their needs. Unless you do want to have **JQ()** and **JQRaw()**.

Thoughts?

**Comment by bitfield** [comment]
Well, it's not just whitespace that needs to be removed from the JQ data; it's quotes, too. A `Strip` method that worked like Ruby's `String#strip` wouldn't do that, I suppose. So we couldn't write this:

```go
script.Echo(data).JQ(query).Strip().Stdout()
```

On the other hand, suppose there were some method `Translate` (`Tr` means nothing unless you already know Unix), that worked like `tr`, we'd have to write:

```go
script.Echo(data).JQ(query).Translate(" \"", "").Stdout()
```

...which is a lot to have to write after every call to `JQ`.

As I say, cleaning up JQ data is such a specific task that I can't see a no-argument method to do it being useful for anything else.

**Comment by kwilczynski** [comment]
Hi @bitfield,

> Well, it's not just whitespace that needs to be removed from the JQ data; it's quotes, too. A `Strip` method that worked like Ruby's `String#strip` wouldn't do that, I suppose. So we couldn't write this:
> 
> ```go
> script.Echo(data).JQ(query).Strip().Stdout()
> ```

The idea to add **Strip()** was not strictly related to **JQ()**, but more as a generic function (a helper, if you wish) that could be used here and elsewhere to remove leading and trailing whitespaces.

> On the other hand, suppose there were some method `Translate` (`Tr` means nothing unless you already know Unix), that worked like `tr`, we'd have to write:
> 
> ```go
> script.Echo(data).JQ(query).Translate(" \"", "").Stdout()
> ```
> 
> ...which is a lot to have to write after every call to `JQ`.

I see your point. OK. To keep it simple and retain the same ergonomics as other functions, introducing **JQRaw()** next to **JQ()** might be a good idea then.

> As I say, cleaning up JQ data is such a specific task that I can't see a no-argument method to do it being useful for anything else.

The **Strip()** and **Translate()** were intended to be as generic as possible so that both could find use in different places too.

**Comment by tjayrush** [comment]
> This could be, but the trim would need to strip spaces AND quotes, and that would make it a weird "Trim" method I guess, WDYT?

The GoLang `string.Trim` routine takes a `cutlist` which can trim multiple different characters. I don't think it would be weird at all.

**Comment by tjayrush** [comment]
> Is there any example you can think of where a 'trim spaces and quotes' method would be used with anything _other_ than `JQ`?

I would use the ability to time not only spaces and quotes, but a lot of other things, in an application like an HTML scraper. I frequently do this -- scrape 1,000 similar pages, strip out data from tables, for example.

**Comment by bitfield** [comment]
Right, but if we had (for example) `JQRaw`, or equivalent, would we really need a general-purpose `Trim` method? What for?

**Comment by tjayrush** [comment]
I would use `Trim` to remove unwanted characters either from the start, the end, or both ends of a string.

**Comment by bitfield** [comment]
Yes, understood, but why would you need to do that? In what real-world application would you need to trim specific characters from either end of a string? Don't take this as a challenge—I'm not saying there's no use case for this—but as an invitation to contribute suggestions for ways this feature might be used.

If a number of intelligent, experienced people cannot, in practice, think of any situation in which they would really use this feature, then I'm happy to leave the feature out. The best tools tend to be the ones that aren't bloated with unnecessary features.

**Comment by phthom** [comment]
Thanks @ John for your great work on script. I am also in the same kind of questions like you guys. On my side I am regularly using jq to quickly strip some json payload. In my case, I am using `jq -r` to use the raw possibility (no quotes around attributes).  For example :   `echo '{ "ip": "192.168.0.2"}' | jq -r .ip` will give `192.168.0.2`. There are a lot of args with the jq and almost the same with the gojq that use in JQ. I try to find a workaround to pass arguments to the JQ implementation (but I didn't succeed) or to define arguments thru environment variables. Did you you try this path ?

**Comment by phthom** [comment]
Yes but you will add a new function to your script list. I am not sure this is what you want ... Also as I mentioned there are many other parameters that I use like the S (slurp) or the -C for instance. But again, after reading the **argument list in gojq,**  this may be **not** a good idea to get access to some parameters. So yes why not **JQr** or JQraw ... Simple and easy. However I also like the idea to have a kinda generic "translator" in your set of functions. Thanks anyway.

**Comment by bitfield** [comment]
It doesn't look like `gojq` has an easy way to get "raw" results, so now I'm leaning towards something like `Trim` for cleaning up the output.

**Comment by bitfield** [comment]
You can do exactly this with `FilterLine` (for example):

```go
script.File(path).JQ(".ip .host").FilterLine(func (s string) string {
  return strings.Trim(s, `"`)
}.String()
```

**Comment by lazysegtree** [comment]
@bitfield Is the conclusion here that we won't be adding any raw flag as it can be done with `FilterLine`

Can you label the issue with `won't do` or something, or maybe rename it to `[Won't do] Add "raw" flag equivalent for JQ`

**Comment by bitfield** [comment]
For the time being, yes.

**Comment by prochac** [comment]
I call for reopen.
FilterLine with strings.Trim works only for JSON strings, but it doesn't solve the nested escaping

```
$ echo '"foo\"bar"' | jq
"foo\"bar"
$ echo '"foo\"bar"' | jq -r 
foo"bar
```

```
$ echo '{"json":"{\"in\":\"json\"}"}'
{"json":"{\"in\":\"json\"}"}
$ echo '{"json":"{\"in\":\"json\"}"}' | jq
{
  "json": "{\"in\":\"json\"}"
}
$ echo '{"json":"{\"in\":\"json\"}"}' | jq '.json'
"{\"in\":\"json\"}"
$ echo '{"json":"{\"in\":\"json\"}"}' | jq -r '.json'
{"in":"json"}
$ echo '{"json":"{\"in\":\"json\"}"}' | jq -r '.json' | jq
{
  "in": "json"
}
$ echo '{"json":"{\"in\":\"json\"}"}' | jq -r '.json' | jq '.in'  
"json"
$ echo '{"json":"{\"in\":\"json\"}"}' | jq -r '.json' | jq -r '.in' 
```

**Comment by lazysegtree** [comment]
@prochac Maybe not via `strings.Trim`, but via `strconv.Unquote`, we can simulate `-r`

<details><summary>Details</summary>
<p>

```bash
➜  examples git:(temp) ✗ go run main.go
echo '{"json":"{\"in\":\"json\"}"}' | jq '.json' ->
"{\"in\":\"json\"}"
echo '{"json":"{\"in\":\"json\"}"}' | jq -r '.json' ->
{"in":"json"}
echo '{"json":"{\"in\":\"json\"}"}' | jq -r '.json' | jq .in' ->
"json"
➜  examples git:(temp) ✗
```

```go
package main

import (
	"fmt"
	"strconv"

	"github.com/bitfield/script"
)

func JQRaw(s string) string {
	res, _ := strconv.Unquote(s)
	return res
}

func main() {
	s := `{"json":"{\"in\":\"json\"}"}`

	fmt.Printf("echo '%s' | jq '.json' ->\n", s)
	_, err := script.Echo(s).JQ(".json").Stdout()
	if err != nil {
		fmt.Printf("err = %v\n", err)
	}
	fmt.Printf("echo '%s' | jq -r '.json' ->\n", s)
	_, err = script.Echo(s).JQ(".json").FilterLine(JQRaw).Stdout()
	if err != nil {
		fmt.Printf("err = %v\n", err)
	}
	fmt.Printf("echo '%s' | jq -r '.json' | jq .in' ->\n", s)
	_, err = script.Echo(s).JQ(".json").FilterLine(JQRaw).JQ(".in").Stdout()
	if err != nil {
		fmt.Printf("err = %v\n", err)
	}
}

```

</p>
</details>

**Comment by bitfield** [comment]
Like I say, script doesn't know anything about JSON. I think let's raise this issue upstream in `gojq` to add the raw behavior. Then we can use it in script.

**Comment by prochac** [comment]
@lazysegtree Thanks.
Unquote seems to be simple and bulletproof. Locally, I was playing with `json.RawMessage`.
